<a href="https://colab.research.google.com/github/Mrnine12/Topicos_A/blob/main/Atividade_II_Topicos_A.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

O conjunto de dados CIFAR-10 (Instituto Canadense de Pesquisa Avançada) é uma coleção de imagens muito utilizada para treinar algoritmos de Machine Learning.

O conjunto de dados possui 60.000 imagens divididas em 10 classes diferentes. As 10 classes diferentes representam aviões, carros, pássaros, gatos, veados, cães, sapos, cavalos, navios e caminhões. Existem 6.000 imagens de cada classe.

**Bibliotecas**

In [6]:
import cv2
import tensorflow as tf
import numpy as np
from keras.datasets import cifar10
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense, Flatten, Conv2D, MaxPooling2D, Dropout
from tensorflow.keras import layers
from keras.utils import to_categorical

In [3]:
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

170500096/170498071 [==============================] - 6s 0us/step


Basicamente, aqui será apresentada uma comparação entre uma rede neural artificial e uma rede neural convolucional, técnica mais indicada para imagens, sendo que ambas as técnicas serão aplicadas ao conjunto de dados CIFAR-10.

## **Artificial Neural Network**

In [7]:
# Grayscale

Xtreino = np.array([cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) for image in x_train])
Xteste = np.array([cv2.cvtColor(image, cv2.COLOR_BGR2GRAY) for image in x_test])

# NORMALIZAR OS VALORES DOS PIXELS

Xtreino = Xtreino / 255
Xteste = Xteste / 255

# TRANSFORMAR NUM VETOR DE DIMENSAO 1

Xtreino = Xtreino.reshape((-1, 1024))
Xteste = Xteste.reshape((-1, 1024))

In [8]:
# ARQUITETURA DA REDE

model_ann = Sequential()

model_ann.add(Dense(64, activation = 'relu', input_dim = 1024))
model_ann.add(Dense(64, activation = 'relu'))
model_ann.add(Dense(10, activation = 'softmax'))

# SETUP DA REDE

model_ann.compile(optimizer = 'adam', loss = 'categorical_crossentropy', metrics = ['accuracy'])

In [9]:
model_ann.fit(Xtreino, to_categorical(y_train), epochs = 20, batch_size = 32)

Epoch 1/20
1563/1563 [==============================] - 3s 2ms/step - loss: 2.1006 - accuracy: 0.2307
Epoch 2/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.9154 - accuracy: 0.3129
Epoch 3/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.8505 - accuracy: 0.3380
Epoch 4/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.8120 - accuracy: 0.3533
Epoch 5/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.7839 - accuracy: 0.3627
Epoch 6/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.7693 - accuracy: 0.3704
Epoch 7/20
1563/1563 [==============================] - 2s 2ms/step - loss: 1.7537 - accuracy: 0.3805
Epoch 8/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.7390 - accuracy: 0.3802
Epoch 9/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.7282 - accuracy: 0.3876
Epoch 10/20
1563/1563 [==============================] - 3s 2ms/step - loss: 1.722

In [10]:
model_ann.evaluate(Xteste, to_categorical(y_test))[1]

313/313 [==============================] - 0s 1ms/step - loss: 1.7109 - accuracy: 0.3919


0.3919000029563904

## **Convolutional Neural Network**

In [11]:
# Criar categorias

y_train_one_hot = to_categorical(y_train)
y_test_one_hot = to_categorical(y_test)

# NORMALIZAR OS VALORES DOS PIXELS

x_train = x_train / 255
x_test = x_test / 255

In [12]:
# Modelo

model = Sequential()

model.add(Conv2D(32, (5, 5), activation = 'relu', input_shape = (32,32,3)))
model.add(MaxPooling2D(pool_size = (2, 2)))
model.add(Conv2D(64, (5, 5), activation = 'relu'))
model.add(MaxPooling2D(pool_size = (2, 2)))
model.add(Flatten())
model.add(Dense(1000, activation = 'relu'))
model.add(Dropout(0.5))
model.add(Dense(500, activation = 'relu'))
model.add(Dropout(0.5))
model.add(Dense(250, activation = 'relu'))
model.add(Dense(10, activation = 'softmax'))

model.compile(loss = 'categorical_crossentropy', optimizer ='adam', metrics = ['accuracy'])

In [13]:
hist = model.fit(x_train, y_train_one_hot, batch_size = 256, epochs = 20, validation_split = 0.2 )

Epoch 1/20
157/157 [==============================] - 69s 438ms/step - loss: 2.0134 - accuracy: 0.2346 - val_loss: 1.4601 - val_accuracy: 0.4690
Epoch 2/20
157/157 [==============================] - 68s 431ms/step - loss: 1.4393 - accuracy: 0.4695 - val_loss: 1.2805 - val_accuracy: 0.5468
Epoch 3/20
157/157 [==============================] - 68s 432ms/step - loss: 1.2499 - accuracy: 0.5514 - val_loss: 1.1598 - val_accuracy: 0.5850
Epoch 4/20
157/157 [==============================] - 68s 434ms/step - loss: 1.1165 - accuracy: 0.6005 - val_loss: 1.0790 - val_accuracy: 0.6184
Epoch 5/20
157/157 [==============================] - 68s 434ms/step - loss: 1.0139 - accuracy: 0.6403 - val_loss: 0.9999 - val_accuracy: 0.6466
Epoch 6/20
157/157 [==============================] - 68s 432ms/step - loss: 0.9337 - accuracy: 0.6690 - val_loss: 0.9616 - val_accuracy: 0.6620
Epoch 7/20
157/157 [==============================] - 68s 433ms/step - loss: 0.8532 - accuracy: 0.6962 - val_loss: 0.9676 - val_ac

In [14]:
model.evaluate(x_test, y_test_one_hot)[1]

313/313 [==============================] - 5s 18ms/step - loss: 1.0486 - accuracy: 0.7041


0.7041000127792358

Ao se comparar os dois modelos, ANN e CNN, foi possivel observar que como esperado, a rede mais propicia para se trabalhar com imagens, a CNN, apresentou resultados superiores ao visto pela ANN. 

Para o modelo convolucional é possivel observar uma acurácia de 70%, enquanto que a acurácia vista para o modelo de rede neural tradicional fica em torno de 39%.

Em ambos os casos, apesar de aumentar o numero de epocas, a acurácia do modelo não apresentou melhoras.